feat!: error handling overhaul, more powerful custom error handlers

m4tx opened the pull request:
This contains a fairly large overhaul of the error handling subsystem. This is the summary of the changes:
- The error can now return proper HTTP status codes.
- It's easier to define custom error pages now.
- Custom error pages now have ways to receive project context, error data, and are affected by middleware (opt-out).
- The error struct is now always 8-bytes large instead of 110+ bytes. The code is optimized for happy path now, and the actual error content is stored on the heap.
- Cot-provided nice, debug-friendly error pages are now only displayed if "Accept: text/html" header was sent. This means the developer won't be flooded with HTML content if testing REST APIs or performing requests outside of the web browser (by using `curl`, for instance).
- Various other reactors.

Fixes #166 
Fixes #338

m4tx commented on cot/src/error/handler.rs:
```diff
@@ -0,0 +1,245 @@
+    note = "make sure the function takes no more than 10 parameters",
+    note = "make sure the function returns a type that implements `IntoResponse`"
+)]
+pub trait ErrorPageHandler<T = ()> {
```
Type safety. We don't have access to the original `request` object, so I don't want to give users a false impression that they can use regular request handlers and extract the same data. Instead, there is a new trait (that's mostly identical to the original), but the extractors are only defined for a handful of original types, so that they won't shoot themselves in the foot as easily.

seqre replied:
Are there any other differences other than using `FromErrorRequestParts` instead of `FromRequestParts` and not using `FromRequest`?

m4tx replied:
No, that's precisely the difference.

m4tx commented on errors.md:
Ahh right, thanks for catching this!

seqre commented on cot/src/project.rs:
```diff
@@ -1783,29 +1873,26 @@ pub async fn run_at_with_shutdown(
             Ok(response) => response,
             Err(error_response) => {
-                if is_debug {
+                if is_debug && allows_html(&request_parts) {
```
My point is about the case of debug disabled and the API request not allowing HTML. In the current state, we'll go to the second branch, which will return a custom/default error page, which will return it to the API request. I suggest to extract `allows_html(&request_parts)` to a separate check before this conditional and if it's empty, don't build any error page.

m4tx replied:
> don't build any error page

So what should we return as a response then? If you need to return JSON, just return it in your custom handler.

m4tx commented on cot/src/project.rs:
```diff
@@ -491,32 +444,35 @@ pub type MiddlewareContext = ProjectContext<WithDatabase>;
 }
 
-impl<S> RootHandlerBuilder<S>
+impl<S, SE> RootHandlerBuilder<S, SE>
 where
     S: Service<Request, Response = Response, Error = Error> + Send + Sync + Clone + 'static,
     S::Future: Send,
+    SE: Service<Request, Response = Response, Error = Error> + Send + Sync + Clone + 'static,
+    SE::Future: Send,
```
Fair enough, let's go with more verbose `RootService` and `ErrorHandlerService`.

m4tx commented:
A summary of the most recent changes:
* Removed the gigantic `cot::error::ErrorRepr` enum, as it didn't scale at all, and by being private, it wasn't particularly useful anyway. Now `cot::Error` is essentially a wrapper over `std::error::Error` (somewhat similarly to `anyhow::Error`)
* Added specific error types for Not Found, Method Not Allowed errors, etc
* Removed the `FromErrorRequestParts` enum as it's no longer needed—`FromRequestHead` is used instead
* Changed all (except form field errors) error descriptions so that they start with a lowercase letter in accordance to Rust's [StdError guidelines](https://doc.rust-lang.org/std/error/trait.Error.html) ("Error messages are typically concise lowercase sentences without trailing punctuation")

m4tx commented on cot/src/project.rs:
```diff
@@ -1262,16 +1294,61 @@ impl Bootstrapper<Initialized> {
     #[must_use]
-    pub fn into_context_and_handler(self) -> (ProjectContext, BoxedHandler) {
-        (self.context, self.handler)
+    pub fn into_context_and_handler(self) -> BootstrappedProject {
```
Yeah, it doesn't make much sense now—let's go with `finish`.

m4tx commented:
For real-world examples of usage, see:
* cot-rs/cot-site#38
* m4tx/chombo-gen#58

seqre commented on cot/src/request/extractors.rs:
```diff
@@ -225,13 +222,17 @@ where
 
+#[derive(Debug, thiserror::Error)]
+#[error("could not parse query parameters: {0}")]
+struct QueryParametersParse(serde_path_to_error::Error<serde::de::value::Error>);
```
```suggestion
struct QueryParametersParseError(serde_path_to_error::Error<serde::de::value::Error>);
```

m4tx commented on cot/src/error.rs:
```diff
@@ -1,304 +1,6 @@
-    }
-}
+pub(crate) mod error_impl;
+pub mod handler;
+pub mod method_not_allowed;
+pub mod not_found;
+pub mod uncaught_panic;
```
Fair enough. What should we do with `not_found::Kind`? Rename to `NotFoundKind`?

seqre replied:
That's a good question, I'm unsure tbh. I think the best solution is to keep it named `Kind`, but re-export as `NotFoundKind`.

m4tx commented on cot/src/error/error_impl.rs:
```diff
@@ -0,0 +1,381 @@
+            }
+            error = inner;
+        }
+        self
```
The idea is that when there's no `cot::Error` with a status code in the chain, we just return the outermost error, as this is the one that will be the most useful. I'll update the docs to reflect that.

m4tx commented on cot/src/error/error_impl.rs:
```diff
@@ -0,0 +1,381 @@
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Display::fmt(&self.inner.inner, f)
+    }
+}
+
+impl StdError for Error {
+    fn source(&self) -> Option<&(dyn StdError + 'static)> {
+        self.inner.inner.source()
+    }
+}
+
+impl Deref for Error {
+    type Target = dyn StdError + Send + Sync;
+
+    fn deref(&self) -> &Self::Target {
+        &*self.inner.inner
+    }
+}
```
First `inner` is `ErrorImpl` instance, and the second is the actual `dyn Error`, so no, we don't skip anything here. I guess this needs some better field names to avoid confusion.

m4tx commented on cot/src/error/error_impl.rs:
```diff
@@ -0,0 +1,381 @@
+    }
+}
+
+impl Debug for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Debug::fmt(&self.inner, f)
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Display::fmt(&self.inner.inner, f)
+    }
+}
```
For `Debug`, we want to include the status code (or lack thereof) in whatever is displayed. For `Display`, we just tell the inner error object to print the message.

seqre commented on cot/src/form.rs:
```diff
@@ -370,13 +372,21 @@ async fn urlencoded_form_data(request: &mut Request) -> Result<Bytes, FormError>
+#[derive(Debug, Error)]
+#[error(
+    "request does not contain a form (expected `application/x-www-form-urlencoded` or \
+        `multipart/form-data` content type, or a GET or HEAD request)"
```
```suggestion
        `multipart/form-data` content type, and is not a GET or HEAD request)"
```

m4tx commented on cot/src/utils/accept_header_parser.rs:
```diff
@@ -0,0 +1,137 @@
+            .split(',')
+            .filter_map(Self::parse_single)
+            .collect();
+        content_types.sort_by(|a, b| {
```
This would be only marginally better given the usual number of elements (no more than 10), and currently at least the order is obvious and deterministic.

m4tx commented on cot/src/project.rs:
```diff
@@ -1864,50 +1990,56 @@ fn build_cot_error_page(
+                ?error,
+                "Error occurred while running custom server error handler; \
+                returning default error page"
+            );
```
It's not a clone; it's just moved into an `Arc`.
